File: leo2-node/src/main/java/org/deku/leo2/node/data/sync/EntityConsumer.java
package org.deku.leo2.node.data.sync;

import com.google.common.base.Stopwatch;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.deku.leo2.messaging.MessagingContext;
import org.deku.leo2.node.data.PersistenceUtil;
import org.deku.leo2.node.data.sync.v1.EntityStateMessage;
import org.deku.leo2.node.data.sync.v1.EntityUpdateMessage;
import org.deku.leo2.node.messaging.ObjectMessageConverter;
import org.springframework.jms.core.JmsTemplate;
import sx.Disposable;

import javax.jms.*;
import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import java.lang.IllegalStateException;
import java.sql.Timestamp;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Created by masc on 18.06.15.
 */
public class EntityConsumer implements Disposable {
    private Log mLog = LogFactory.getLog(this.getClass());

    private static final int RECEIVE_TIMEOUT = 5000;
    /** Messaging context */
    private MessagingContext mMessagingContext;
    /** Entity manager factory */
    private EntityManagerFactory mEntityManagerFactory;
    /** Spring jms communication abstraction */
    private JmsTemplate mTemplate;
    private ObjectMessageConverter mObjectMessageConverter = new ObjectMessageConverter(
            ObjectMessageConverter.SerializationType.KRYO,
            ObjectMessageConverter.CompressionType.GZIP);

    ExecutorService mExecutorService;

    /**
     * c'tor
     * @param messagingContext
     */
    public EntityConsumer(MessagingContext messagingContext, EntityManagerFactory entityManagerFactory) {
        mMessagingContext = messagingContext;
        mEntityManagerFactory = entityManagerFactory;
        mExecutorService = Executors.newSingleThreadExecutor();
    }

    /**
     * Request entity update
     * @param entityType
     */
    public void request(Class entityType) {
        mExecutorService.submit(() -> {
            try {
                mObjectMessageConverter.resetStatistics();

                EntityManager em = mEntityManagerFactory.createEntityManager();
                EntityRepository er = new EntityRepository(em, entityType);

                Timestamp timestamp = er.findMaxTimestamp();

                Connection cn = mMessagingContext.getConnectionFactory().createConnection();
                cn.start();
                Session session = cn.createSession(false, Session.AUTO_ACKNOWLEDGE);

                Stopwatch sw = Stopwatch.createStarted();

                Queue requestQueue = session.createQueue(EntityStateMessage.ENTITY_QUEUE_NAME);
                TemporaryQueue receiveQueue = session.createTemporaryQueue();

                mLog.info(String.format("Requesting entities of type [%s]", entityType.toString()));

                // Send entity state message
                MessageProducer mp = session.createProducer(requestQueue);
                mp.setDeliveryMode(DeliveryMode.NON_PERSISTENT);
                mp.setTimeToLive(5 * 60 * 1000);
                Message msg = mObjectMessageConverter.toMessage(
                        new EntityStateMessage(entityType, timestamp),
                        session);
                msg.setJMSReplyTo(receiveQueue);
                mp.send(msg);

                // Receive entity update message
                MessageConsumer mc = session.createConsumer(receiveQueue);
                msg = mc.receive(RECEIVE_TIMEOUT);
                if (msg == null)
                    throw new TimeoutException("Timeout while waiting for entity update message");
                EntityUpdateMessage euMessage = (EntityUpdateMessage) mObjectMessageConverter.fromMessage(msg);

                mLog.debug(euMessage);
                AtomicLong count = new AtomicLong();

                // Transaction processing thread pool
                ExecutorService executorService = Executors.newFixedThreadPool(
                        Runtime.getRuntime().availableProcessors());

                if (euMessage.getAmount() > 0) {
                    PersistenceUtil.transaction(em, () -> {
                        if (!er.hasTimestampAttribute()) {
                            mLog.debug("No timestamp attribute found -> removing all entities");
                            er.removeAll();
                        }

                        // Receive entities
                        boolean eos = false;
                        long lastJmsTimestamp = 0;
                        do {
                            Message tMsg = mc.receive(RECEIVE_TIMEOUT);
                            if (tMsg == null)
                                throw new TimeoutException("Timeout while waiting for next entities chunk");

                            // Verify message order
                            if (tMsg.getJMSTimestamp() < lastJmsTimestamp)
                                throw new IllegalStateException(
                                        String.format("Inconsistent message order (%d < %d)", tMsg.getJMSTimestamp(), timestamp));

                            // Store last timestamp
                            lastJmsTimestamp = tMsg.getJMSTimestamp();

                            eos = tMsg.propertyExists(EntityUpdateMessage.EOS_PROPERTY);

                            if (!eos) {
                                // Deserialize entities
                                List entities = Arrays.asList((Object[]) mObjectMessageConverter.fromMessage(tMsg));

                                // TODO: exceptions within transactions behave in a strange way.
                                // data of transactions that were committed may not be there and h2 may report
                                // cache level state nio exceptions.
                                // Data seems to remain consistent though

                                if (timestamp != null) {
                                    mLog.trace("Removing existing entities");
                                    // If there's already entities, clean out existing first.
                                    // it's much faster than merging everything
                                    for (Object o : entities) {
                                        Object o2 = em.find(entityType,
                                                em.getEntityManagerFactory().getPersistenceUnitUtil().getIdentifier(o));

                                        if (o2 != null) {
                                            em.remove(o2);
                                        }
                                    }
                                    em.flush();
                                    em.clear();
                                }

                                // Persist entities
                                for (Object o : entities) {
                                    em.persist(o);
                                }
                                em.flush();
                                em.clear();
                                count.getAndUpdate(c -> c + entities.size());
                            }
                        } while (!eos);
                    });

                }
                mLog.trace("Joining transaction threads");
                executorService.shutdown();
                executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
                mLog.info(String.format("Received and stored %d in %s (%d bytes)", count.get(), sw.toString(), mObjectMessageConverter.getBytesRead()));

                em.close();
            } catch (TimeoutException e) {
                mLog.error(e.getMessage());
            } catch (Exception e) {
                mLog.error(e.getMessage(), e);
            }

            return null;
        });
    }

    @Override
    public void dispose() {
        mExecutorService.shutdown();
        try {
            mExecutorService.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
        }
    }
}

File: leoz-node/src/main/java/org/deku/leoz/node/service/internal/sync/EntityConsumer.kt
package org.deku.leoz.node.service.internal.sync

import org.deku.leoz.node.data.repository.SyncRepository
import org.deku.leoz.node.service.internal.sync.EntityUpdateMessage.Companion.EOS_PROPERTY
import sx.log.slf4j.debug
import sx.log.slf4j.info
import sx.mq.MqChannel
import sx.mq.MqHandler
import sx.mq.jms.JmsEndpoint
import sx.mq.jms.channel
import sx.mq.jms.listeners.SpringJmsListener
import sx.persistence.transaction
import sx.persistence.truncate
import java.util.concurrent.Executor
import java.util.concurrent.ExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import javax.persistence.EntityManagerFactory

/**
 * Entity consumer
 * Created by masc on 18.06.15.
 * @param messagingConfiguration
 */
class EntityConsumer(
        /** Messaging context  */
        private val requestEndpoint: JmsEndpoint,
        private val notificationEndpoint: JmsEndpoint,

        /** Entity manager factory  */
        private val entityManagerFactory: EntityManagerFactory,
        /** Sync repository */
        private val syncRepository: SyncRepository,

        /** Executor used for listening/processing incoming messages */
        listenerExecutor: Executor,

        private val presets: List<ConsumerPreset<*>>
) :
        SpringJmsListener(notificationEndpoint, listenerExecutor),
        MqHandler<EntityStateMessage> {

    /**
     * Executor for internal tasks, eg issuing requests
     */
    private var executorService: ExecutorService

    private val presetByType: Map<Class<*>, ConsumerPreset<*>>

    init {
        this.addDelegate(this)

        this.presetByType = presets.associateBy { it.entityPath.type }

        // Use single-threaded executor for sequential processing
        this.executorService = java.util.concurrent.Executors.newSingleThreadExecutor { r ->
            val t = Thread(r)
            t.priority = Thread.MIN_PRIORITY
            t
        }
    }

    /**
     * Entity state message handler
     */
    override fun onMessage(message: EntityStateMessage, replyChannel: MqChannel?) {
        this.request(message.entityType!!, message.syncId)
    }

    val requestClient by lazy {
        this.requestEndpoint.channel()
    }

    /**
     * Request entity update
     * @param entityType Entity type
     * @param remoteSyncId Optional remote sync-id, usually provided via notification
     * @param requestAll Ignores sync id, requests all entities (mainly for testing)
     * @param clean Clears all data before requesting
     */
    @Synchronized
    fun request(
            entityType: Class<*>,
            remoteSyncId: Long? = null,
            requestAll: Boolean = false,
            clean: Boolean = false) {

        val preset = presetByType.get(entityType)

        if (preset == null) {
            log.warn("No consumer preset for type [${entityType}]")
            return
        }

        executorService.submit<Unit> {
            // Log formatting with entity type
            val lfmt = { s: String -> "[" + entityType.canonicalName + "]" + " " + s }

            var em: javax.persistence.EntityManager? = null
            try {
                em = entityManagerFactory.createEntityManager()

                if (clean) {
                    em.transaction {
                        em.truncate(entityType)
                    }
                }

                var syncIds: LongRange? = null

                if (!requestAll) {
                    if (preset.syncIdPath != null) {
                        syncIds = syncRepository.findSyncIdMinMax(
                                preset.entityPath,
                                preset.syncIdPath
                        )
                    }

                    if (syncIds != null && remoteSyncId != null && remoteSyncId <= syncIds.endInclusive) {
                        log.debug(lfmt("up-to-date"))
                        return@submit
                    }
                }

                if (preset.syncIdPath == null) {
                    log.debug { lfmt("No sync-id -> remove all entities") }
                    em.transaction {
                        em.truncate(entityType)
                    }
                }

                val sw = com.google.common.base.Stopwatch.createStarted()

                // Send entity state message
                this.requestClient.createReplyClient().use { replyClient ->
                    var totalCount: Long = 0
                    var bytesReceived: Long = 0

                    while (true) {
                        var count = 0
                        log.info { lfmt("Requesting with sync-id [${syncIds?.endInclusive}]") }

                        requestClient.sendRequest(
                                message = EntityStateMessage(entityType = entityType, syncId = syncIds?.endInclusive),
                                replyChannel = replyClient)

                        // Receive entity update message
                        val euMessage = replyClient.receive(
                                EntityUpdateMessage::class.java)

                        log.debug { lfmt("${euMessage}") }

                        if (euMessage.amount == 0L)
                            break

                        val emv = em!!
                        em.transaction {
                            // Receive entities
                            var eos: Boolean
                            var lastJmsTimestamp: Long = 0
                            do {
                                val tMsg = replyClient.receive()
                                        ?: throw TimeoutException("Timeout while waiting for next entities chunk")

                                // Verify message order
                                if (tMsg.jmsTimestamp < lastJmsTimestamp)
                                    throw IllegalStateException(
                                            String.format("Inconsistent message order (%d < %d)", tMsg.jmsTimestamp, syncIds?.endInclusive))

                                // Store last timestamp
                                lastJmsTimestamp = tMsg.jmsTimestamp

                                eos = tMsg.propertyExists(EOS_PROPERTY)

                                if (!eos) {
                                    // Deserialize entities
                                    val entities = replyClient.converter.fromMessage(tMsg, { size ->
                                        bytesReceived += size
                                    }) as Array<*>

                                    if (syncIds != null || requestAll) {
                                        // If there's already entities, clean out existing first.
                                        // it's much faster than merging everything
                                        for (o in entities) {
                                            val o2 = emv.find(entityType,
                                                    emv.entityManagerFactory.persistenceUnitUtil.getIdentifier(o))

                                            if (o2 != null) {
                                                emv.remove(o2)
                                            }
                                        }
                                        emv.flush()
                                        emv.clear()
                                    }

                                    // Persist entities
                                    for (o in entities) {
                                        emv.persist(o)
                                    }
                                    emv.flush()
                                    emv.clear()

                                    count += entities.size
                                }
                            } while (!eos)
                        }

                        totalCount += count

                        // For sync-id-less syncs, batches cannot be supported (TODO: remove support for this)
                        if (syncIds == null)
                            break

                        if (count < euMessage.batchSize)
                            break

                        syncIds = if (preset.syncIdPath != null) {
                            syncRepository.findSyncIdMinMax(
                                    preset.entityPath,
                                    preset.syncIdPath
                            )
                        } else null
                    }

                    log.info { lfmt("Received and stored ${totalCount} in ${sw} (${bytesReceived} bytes)") }
                }
            } catch (e: TimeoutException) {
                log.error(lfmt(e.message ?: ""))
            } catch (e: Throwable) {
                log.error(lfmt(e.message ?: ""), e)
            } finally {
                if (em != null)
                    em.close()
            }
        }
    }

    /**
     * Request entity update
     * @param entityType Entity type
     */
    fun request(entityType: Class<*>, clean: Boolean = false) {
        this.request(
                entityType = entityType,
                remoteSyncId = null,
                clean = clean)
    }

    /**
     * Request entity updates for all presets
     */
    fun request(clean: Boolean = false) {
        this.presets.forEach {
            this.request(
                    entityType = it.entityPath.type,
                    clean = clean
            )
        }
    }

    override fun close() {
        executorService.shutdownNow()
        try {
            executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
            // Interruption is ok.
        }

        super.close()
    }
}
